### PR TITLE
Add compatibility with enterprise-wide runners

### LINKED ISSUE #45: Support to register runner at enterprise account level
Please support the new feature of having runners at enterprise account level:
https://github.blog/changelog/2020-08-05-github-actions-enterprise-runners-and-runner-groups

### PR DESCRIPTION
**Pull request by jakubstefan**
Resolves #45, related to #95.

Adds the possibility to register the runner enterprise-wide. Also, as having now multiple scopes, I considered organising a bit the selection of the scope.

Tested in Enterprise environment.

### CODE REVIEW

**Comment by retzero** [comment]
It also works in our GitHub Ent. Thank you for the changes!

**Comment by jakubstefan** [comment]
@myoung34 How should we proceed with this? Do you need any more information to support the changes?

**Comment by myoung34** [comment]
Sorry, haven't had a chance to review. I'm out this weekend but will review it Monday

**Comment by jakubstefan** [comment]
No worries. Have a nice weekend.

File: entrypoint.sh [hunk 1]
```diff
@@ -17,13 +17,32 @@ _RUNNER_NAME=${RUNNER_NAME:-${RUNNER_NAME_PREFIX:-github-runner}-$(head /dev/ura
 _RUNNER_WORKDIR=${RUNNER_WORKDIR:-/_work}
 _LABELS=${LABELS:-default}
 _RUNNER_GROUP=${RUNNER_GROUP:-Default}
-_SHORT_URL=${REPO_URL}
 _GITHUB_HOST=${GITHUB_HOST:="github.com"}
 
-if [[ ${ORG_RUNNER} == "true" ]]; then
```
Comment (myoung34) [inline entrypoint.sh:23]: For now we should find a way to make it work as-is (with ORG_RUNNER) and as new (RUNNER_SCOPE )

There are quite a few users of this and I'd like to throw some deprecation warnings + a timeline before making a hard change.

File: entrypoint.sh [hunk 2]
```diff
@@ -17,13 +17,32 @@ _RUNNER_NAME=${RUNNER_NAME:-${RUNNER_NAME_PREFIX:-github-runner}-$(head /dev/ura
...
+if [[ -z $RUNNER_SCOPE ]]; then
+  if [[ ${ORG_RUNNER} == "true" ]]; then
+    export RUNNER_SCOPE="org"
+  else
+    export RUNNER_SCOPE="repo"
+  fi
 fi
 
+case ${RUNNER_SCOPE} in
+  "org")
```
Comment (myoung34) [inline entrypoint.sh]: for string based we should make it case insensitive and allow shortening for things like org|organization

File: token.sh [hunk 3]
```diff
@@ -14,22 +13,37 @@ API_VERSION=v3
 API_HEADER="Accept: application/vnd.github.${API_VERSION}+json"
 AUTH_HEADER="Authorization: token ${ACCESS_TOKEN}"
 
-REPO_URL=${REPO_URL:-${URI}}
+if [[ ${RUNNER_SCOPE} == "repo" ]]; then
+  [[ -z ${REPO_URL} ]] && ( echo "REPO_URL required for repo runners"; exit 1 )
```
Comment (myoung34) [inline token.sh]: this already exists in `entrypoint.sh` and shouldnt be a concern in both places (keep it DRY)

File: token.sh [hunk 4]
```diff
@@ -14,22 +13,37 @@ API_VERSION=v3
...
 
-_FULL_URL="${URI}/repos/${_ACCOUNT}/${_REPO}/actions/runners/registration-token"
-if [[ ${_ORG_RUNNER} == "true" ]]; then
-  [[ -z ${ORG_NAME} ]] && ( echo "ORG_NAME required for org runners"; exit 1 )
-  _FULL_URL="${URI}/orgs/${ORG_NAME}/actions/runners/registration-token"
-  _SHORT_URL="${_PROTO}${_GITHUB_HOST}/${ORG_NAME}"
-else
-  _SHORT_URL=$REPO_URL
-fi
+case ${RUNNER_SCOPE} in
```
Comment (myoung34) [inline token.sh:16]: id like to see if theres a way to shorten the next bit, it seems repetitive from `entrypoint.sh`

**Review by myoung34 — CHANGES_REQUESTED** [review]

**Comment by jakubstefan** [comment]
I have done some changes to be more flexible with the `RUNNER_SCOPE` value. I have done some code cleaning as well.

About the concerns of backwards compatibility, it ensures it. You can continue using `ORG_RUNNER` without any issue. The only comment is that `RUNNER_SCOPE` is stronger. Maybe you would like to keep `ORG_RUNNER` in the readme?

Here you can find the behaviour for different cases:

<details><summary>Tests</summary>

```
WITH:
    GITHUB_HOST=my-host.com
    RUNNER_SCOPE=enterprise
    ENTERPRISE_NAME=my-enterprise
RESULT:
    RUNNER_SCOPE: enterprise
    _SHORT_URL: https://my-host.com/enterprises/my-enterprise
    _FULL_URL: https://my-host.com/api/v3/enterprises/my-enterprise/actions/runners/registration-token

WITH:
    GITHUB_HOST=my-host.com
    RUNNER_SCOPE=ent
    ENTERPRISE_NAME=my-enterprise
RESULT:
    RUNNER_SCOPE: enterprise
    _SHORT_URL: https://my-host.com/enterprises/my-enterprise
    _FULL_URL: https://my-host.com/api/v3/enterprises/my-enterprise/actions/runners/registration-token

WITH:
    GITHUB_HOST=my-host.com
    RUNNER_SCOPE=org
    ORG_NAME=my-org
RESULT:
    RUNNER_SCOPE: org
    _SHORT_URL: https://my-host.com/my-org
    _FULL_URL: https://my-host.com/api/v3/orgs/my-org/actions/runners/registration-token

WITH:
    GITHUB_HOST=my-host.com
    RUNNER_SCOPE=organisation
    ORG_NAME=my-org
RESULT:
    RUNNER_SCOPE: org
    _SHORT_URL: https://my-host.com/my-org
    _FULL_URL: https://my-host.com/api/v3/orgs/my-org/actions/runners/registration-token

WITH:
    GITHUB_HOST=my-host.com
    RUNNER_SCOPE=organization
    ORG_NAME=my-org
RESULT:
    RUNNER_SCOPE: org
    _SHORT_URL: https://my-host.com/my-org
    _FULL_URL: https://my-host.com/api/v3/orgs/my-org/actions/runners/registration-token

WITH:
    GITHUB_HOST=my-host.com
    RUNNER_SCOPE=ORG
    ORG_NAME=my-org
RESULT:
    RUNNER_SCOPE: org
    _SHORT_URL: https://my-host.com/my-org
    _FULL_URL: https://my-host.com/api/v3/orgs/my-org/actions/runners/registration-token

WITH:
    GITHUB_HOST=my-host.com
    RUNNER_SCOPE=repo
    REPO_URL=https://my-host.com/my_org/my_repo
RESULT:
    RUNNER_SCOPE: repo
    _SHORT_URL: https://my-host.com/my_org/my_repo
    _FULL_URL: https://my-host.com/api/v3/repos/my_org/my_repo/actions/runners/registration-token

WITH:
    GITHUB_HOST=my-host.com
    ORG_RUNNER=true
    ORG_NAME=my-org
RESULT:
    ORG_RUNNER: true
    RUNNER_SCOPE: org
    _SHORT_URL: https://git.my-host.com/my-org
    _FULL_URL: https://git.my-host.com/api/v3/orgs/my-org/actions/runners/registration-token

WITH:
    GITHUB_HOST=my-host.com
    ORG_NAME=my-org
    REPO_URL=https://my-host.com/my_org/my_repo
RESULT:
    RUNNER_SCOPE: repo
    _SHORT_URL: https://my-host.com/my_org/my_repo
    _FULL_URL: https://my-host.com/api/v3/repos/my_org/my_repo/actions/runners/registration-token

WITH:
    GITHUB_HOST=my-host.com
    RUNNER_SCOPE=repo
    ORG_RUNNER=true
    ORG_NAME=my-org
    REPO_URL=https://my-host.com/my_org/my_repo
RESULT:
    ORG_RUNNER: true
    RUNNER_SCOPE: repo
    _SHORT_URL: https://my-host.com/my_org/my_repo
    _FULL_URL: https://my-host.com/api/v3/repos/my_org/my_repo/actions/runners/registration-token
```

</details>

**Comment by myoung34** [comment]
Looks great!

You can remove the bits in the README about `ORG_RUNNER` and ill work a follow up PR to deprecate it

**Comment by myoung34** [comment]
Those tests are exactly what I was hoping for!

I'll merge it and modify the README so you can get this closed out

Just note: I don't backport releases, so this will be in `latest` only until upstream actions cuts a  new release which auto builds here

**Review by myoung34 — APPROVED** [review]

**Comment by jakubstefan** [comment]
Thanks for the review :slightly_smiling_face: !

**Pull request merged**